File: 0983) Minimum Cost For Tickets.cpp
class Solution {
public:
    int mincostTickets(vector<int>& days, vector<int>& costs) {
        int n = days.size();
        vector<int> dp(n+1, -1);
        return helper(days, costs, 0, dp);
    }
    int helper(vector<int> &days, vector<int> &costs, int id, vector<int> &dp)
    {
        int n = days.size();
        if(id >= n)
        {
            return 0;
        }
        if(dp[id] != -1)
        {
            return dp[id];
        }
        int cost1 = costs[0] + helper(days, costs, id+1, dp);
        int id1 = id;
        while(id1 < n && days[id1] < days[id] + 7)
        {
            id1++;
        }
        int cost2 = costs[1] + helper(days, costs, id1, dp);
        int id2 = id;
        while(id2 < n && days[id2] < days[id] + 30)
        {
            id2++;
        }
        int cost3 = costs[2] + helper(days, costs, id2, dp);
        dp[id] = min({cost1, cost2, cost3});
        return dp[id];
    }
};

File: 0983) Minimum Cost For Tickets.swift
class Solution {
    func mincostTickets(_ days: [Int], _ costs: [Int]) -> Int {
        let lastDay = days.last!
         var dp = Array(repeating: 0 ,count: lastDay+1)
         var travelDays = Set(days)
         for day in 1...lastDay{
            if !travelDays.contains(day){
                dp[day]=dp[day-1]
            }else{
                let const1Day=dp[max(0, day-1)]+costs[0]
                let const7Day=dp[max(0, day-7)]+costs[1]
                let const30Day=dp[max(0, day-30)]+costs[2]
                dp[day]=min(const1Day, const7Day ,const30Day)
                
            }
         }
         return dp [lastDay]
    }
}
